src/nci_util.c

```diff
@@ -849,6 +850,114 @@ nci_parse_intf_activated_ntf(
     return FALSE;
 }
 
+static
+gsize
+nci_mode_param_copy_impl(
+    NciModeParam* dest,
+    const NciModeParam* src,
+    NCI_MODE mode)
+{
+    if (G_LIKELY(src) && G_LIKELY(dest)) {
+        gsize size = sizeof(NciModeParam);
+        const NciModeParamPollB* poll_b = NULL;
+        const NciModeParamListenF* listen_f = NULL;
+
+        switch (mode) {
+        case NCI_MODE_ACTIVE_POLL_A:        /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_A:
+        case NCI_MODE_ACTIVE_POLL_F:
+        case NCI_MODE_PASSIVE_POLL_F:
+            memcpy(dest, src, size);
+            return size;
+        case NCI_MODE_PASSIVE_POLL_B:
+            poll_b = &src->poll_b;
+            memcpy(dest, src, size);
+            if (poll_b->prot_info.bytes) {
+                size = G_ALIGN8(size);
+                guint8* ptr = (guint8*)dest + size;
+
+                memcpy(ptr, poll_b->prot_info.bytes, poll_b->prot_info.size);
+                dest->poll_b.prot_info.bytes = ptr;
+                size += G_ALIGN8(poll_b->prot_info.size);
+            }
+            return size;
+        case NCI_MODE_ACTIVE_LISTEN_F:
+        case NCI_MODE_PASSIVE_LISTEN_F:
+            listen_f = &src->listen_f;
+            memcpy(dest, src, size);
+            if (listen_f->nfcid2.bytes) {
+                size = G_ALIGN8(size);
+                guint8* ptr = (guint8*)dest + size;
+
+                memcpy(ptr, listen_f->nfcid2.bytes, listen_f->nfcid2.size);
+                dest->listen_f.nfcid2.bytes = ptr;
+                size += G_ALIGN8(listen_f->nfcid2.size);
+            }
+            return size;
+        case NCI_MODE_PASSIVE_POLL_15693:   /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_15693:
+            break;
+        case NCI_MODE_PASSIVE_LISTEN_A:     /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_B:
+        case NCI_MODE_ACTIVE_LISTEN_A:
+            /* NCI 1.0 defines no parameters for A/B Listen modes */
+            return 0;
+        }
+        GDEBUG("Unhandled activation mode 0x%02x", mode);
+        return 0;
+    } else {
+        return 0;
+    }
```

Comment:
How about single `return 0;` at the end?

Reply:
Thanks, I have fixed that